File: cpp_component/component/hub/gatemanager.cpp
/*
 * qianqians
 * 2020-1-10
 * gatemanager.cpp
 */
#include <omp.h>

#include <enetacceptservice.h>
#include <redismqservice.h>
#include <modulemng_handle.h>

#include "gatemanager.h"
#include "hub_service.h"

namespace hub {
	
gateproxy::gateproxy(std::shared_ptr<abelkhan::Ichannel> ch, std::shared_ptr<hub_service> hub, std::string gate_name) {
	_ch = ch;
	_hub_call_gate_caller = std::make_shared<abelkhan::hub_call_gate_caller>(ch, service::_modulemng);
	_hub = hub;
	_gate_name = gate_name;
}

void gateproxy::reg_hub() {
	_hub_call_gate_caller->reg_hub(_hub->name_info.name, _hub->hub_type)->callBack([this]() {
		spdlog::trace("hub reg_hub to gate:{0} sucessed", _gate_name);
		sig_reg_hub_sucessed.emit();
	}, [this]() {
		spdlog::trace("hub reg_hub to gate:{0} faild", _gate_name);
	})->timeout(5 * 1000, [this]() {
		spdlog::trace("hub reg_hub to gate:{0} timeout", _gate_name);
	});
}

std::shared_ptr<abelkhan::hub_call_gate_reverse_reg_client_hub_cb> gateproxy::reverse_reg_client_hub(std::string client_uuid) {
	return _hub_call_gate_caller->reverse_reg_client_hub(client_uuid);
}

void gateproxy::unreg_client_hub(std::string client_uuid) {
	_hub_call_gate_caller->unreg_client_hub(client_uuid);
}

void gateproxy::disconnect_client(std::string& cuuid) {
	_hub_call_gate_caller->disconnect_client(cuuid);
}

void gateproxy::forward_hub_call_client(const std::string& cuuid, const std::vector<uint8_t>& _data_bin) {
	_hub_call_gate_caller->forward_hub_call_client(cuuid, _data_bin);
}

void gateproxy::forward_hub_call_group_client(const std::vector<std::string>& cuuids, const std::vector<uint8_t>& rpc_argv) {
	_hub_call_gate_caller->forward_hub_call_group_client(cuuids, rpc_argv);
}

void gateproxy::forward_hub_call_global_client(const std::vector<uint8_t>& rpc_argv) {
	_hub_call_gate_caller->forward_hub_call_global_client(rpc_argv);
}

directproxy::directproxy(std::string cuuid_, std::shared_ptr<abelkhan::Ichannel> direct_ch) {
	_hub_call_client_caller = std::make_shared<abelkhan::hub_call_client_caller>(direct_ch, service::_modulemng);
	
	_cuuid = cuuid_;
	_direct_ch = direct_ch;
}

void directproxy::call_client(const std::vector<uint8_t>& rpc_argv) {
	_hub_call_client_caller->call_client(rpc_argv);
}

gatemanager::gatemanager(std::shared_ptr<service::enetacceptservice> conn_, std::shared_ptr<hub_service> hub_) {
	_conn_enet = conn_;
	_hub = hub_;

	Init();
}

gatemanager::gatemanager(std::shared_ptr<service::redismqservice> conn_, std::shared_ptr<hub_service> hub_) {
	_conn_redismq = conn_;
	_hub = hub_;

	Init();
}

void gatemanager::Init() {
	_hub->sig_svr_be_closed.connect([this](std::string svr_type, std::string svr_name) {
		if (svr_type != "gate") {
			return;
		}

		auto old_it = wait_destory_gates.find(svr_name);
		if (old_it != wait_destory_gates.end()) {
			for (auto it = clients.begin(); it != clients.end(); ) {
				if (it->second == old_it->second) {
					it = clients.erase(it);
				}
				else {
					it++;
				}
			}

			ch_gates.erase(old_it->second->_ch);
			wait_destory_gates.erase(old_it);
		}
		else {
			for (auto it = clients.begin(); it != clients.end(); ) {
				if (it->second->_gate_name == svr_name) {
					it = clients.erase(it);
				}
				else {
					it++;
				}
			}

			auto it = gates.find(svr_name);
			if (it != gates.end()) {
				ch_gates.erase(it->second->_ch);
				gates.erase(it);

				_hub->sig_gate_closed.emit(svr_name);
			}
		}
	});
}

void gatemanager::connect_gate(std::string gate_name, std::string host, uint16_t port) {
	spdlog::trace("connect_gate host:{0} port:{1}", host, port);
	auto ip = service::DNS(host);
	_conn_enet->connect(ip, port, [this, host, port, gate_name](std::shared_ptr<abelkhan::Ichannel> ch) {
		spdlog::trace("gate host:{0}  port:{1} reg_hub", host, port);
		
		auto _gate_proxy = std::make_shared<gateproxy>(ch, _hub, gate_name);
		_gate_proxy->sig_reg_hub_sucessed.connect([this, gate_name, ch, _gate_proxy]() {

			auto it = gates.find(gate_name);
			if (it != gates.end()) {
				wait_destory_gates.insert(std::make_pair(it->first, it->second));
			}

			gates[gate_name] = _gate_proxy;
			ch_gates[ch] = _gate_proxy;
		});
		_gate_proxy->reg_hub();
	});
}

void gatemanager::connect_gate(std::string gate_name) {
	spdlog::trace("connect_gate name:{0}", gate_name);
	auto ch = _conn_redismq->connect(gate_name);
	auto _gate_proxy = std::make_shared<gateproxy>(ch, _hub, gate_name);
	_gate_proxy->sig_reg_hub_sucessed.connect([this, gate_name, ch, _gate_proxy]() {

		auto it = gates.find(gate_name);
		if (it != gates.end()) {
			wait_destory_gates.insert(std::make_pair(it->first, it->second));
		}

		gates[gate_name] = _gate_proxy;
		ch_gates[ch] = _gate_proxy;
	});
	_gate_proxy->reg_hub();
}

void gatemanager::client_connect(std::string client_uuid, std::shared_ptr<abelkhan::Ichannel> gate_ch) {
	auto it = ch_gates.find(gate_ch);
	if (it == ch_gates.end()) {
		spdlog::error("unreg gate channel!");
		return;
	}

	spdlog::trace("reg client:{0}", client_uuid);
	clients[client_uuid] = it->second;
}

std::shared_ptr<gateproxy> gatemanager::get_client_gate(std::string client_uuid) {
	auto it = clients.find(client_uuid);
	if (it != clients.end()) {
		return it->second;
	}
	return nullptr;
}

std::shared_ptr<gateproxy> gatemanager::client_seep(std::string client_uuid, std::string gate_name) {
	auto it = gates.find(gate_name);
	if (it == gates.end()) {
		spdlog::trace("unreg gate name:{0}!", gate_name);
		return nullptr;
	}

	spdlog::trace("client_seep client:{0}", client_uuid);
	clients[client_uuid] = it->second;

	return it->second;
}

void gatemanager::client_disconnect(std::string client_uuid) {
	if (clients.find(client_uuid) == clients.end()) {
		return;
	}

	clients.erase(client_uuid);
	_hub->sig_client_disconnect.emit(client_uuid);
}

void gatemanager::client_direct_connect(std::string client_uuid, std::shared_ptr<abelkhan::Ichannel> direct_ch) {
	auto it = direct_clients.find(client_uuid);
	if (it != direct_clients.end()) {
		ch_direct_clients.erase(it->second->_direct_ch);
		direct_clients.erase(it);
	}

	spdlog::trace("reg direct client:{0}", client_uuid);

	auto _directproxy = _directproxy_pool.make_obj(client_uuid, direct_ch);
	direct_clients.insert(std::make_pair(client_uuid, _directproxy));

	ch_direct_clients.insert(std::make_pair(direct_ch, _directproxy));
}

void gatemanager::client_direct_disconnect(std::shared_ptr<abelkhan::Ichannel> direct_ch) {
	auto it = ch_direct_clients.find(direct_ch);
	if (it == ch_direct_clients.end()) {
		return;
	}

	auto cuuid = it->second->_cuuid;
	_directproxy_pool.recycle(it->second);

	direct_clients.erase(cuuid);
	ch_direct_clients.erase(it);

	_hub->sig_direct_client_disconnect.emit(cuuid);


	spdlog::trace("client_direct_disconnect direct_clients.size:{0}, ch_direct_clients.size:{1}!", direct_clients.size(), ch_direct_clients.size());
}

std::shared_ptr<directproxy> gatemanager::get_direct_client(std::shared_ptr<abelkhan::Ichannel> direct_ch) {
	auto it = ch_direct_clients.find(direct_ch);
	if (it == ch_direct_clients.end()) {
		return nullptr;
	}

	return it->second;
}

void gatemanager::disconnect_client(std::string uuid) {
	auto it_gate = clients.find(uuid);
	if (it_gate != clients.end()) {
		it_gate->second->disconnect_client(uuid);
		clients.erase(it_gate);
	}

	auto it_direct = direct_clients.find(uuid);
	if (it_direct != direct_clients.end()) {
		it_direct->second->_direct_ch->disconnect();
		ch_direct_clients.erase(it_direct->second->_direct_ch);
		direct_clients.erase(it_direct);
	}
}

void gatemanager::heartbeat_client(int64_t ticktime) {
	std::vector<std::shared_ptr<directproxy> > remove_client;
	std::vector<std::shared_ptr<directproxy> > exception_client;
	for (auto item : direct_clients) {
		auto proxy = item.second;
		if (proxy->_timetmp > 0 && (proxy->_timetmp + 10 * 1000) < ticktime) {
			remove_client.push_back(proxy);
		}
		if (proxy->_timetmp > 0 && proxy->_theory_timetmp > 0 && (proxy->_theory_timetmp - proxy->_timetmp) > 10 * 1000) {
			exception_client.push_back(proxy);
		}
	}

	for (auto _client : remove_client) {
		client_direct_disconnect(_client->_direct_ch);
	}

	for (auto _client : exception_client) {
		_hub->sig_client_exception.emit(_client->_cuuid);
	}
}

void gatemanager::call_client(const std::string& cuuid, const std::string& func, const msgpack11::MsgPack::array& argvs) {
	msgpack11::MsgPack::array event_;
	event_.push_back(func);
	event_.push_back(argvs);
	msgpack11::MsgPack _pack(event_);
	auto data = _pack.dump();
	std::vector<uint8_t> _data_bin;
	_data_bin.resize(data.size());
	memcpy(_data_bin.data(), data.data(), data.size());
	
	auto it_direct_client = direct_clients.find(cuuid);
	if (it_direct_client != direct_clients.end()) {
		it_direct_client->second->call_client(_data_bin);
		return;
	}

	if (clients.find(cuuid) == clients.end()) {
		spdlog::error("unreg client cuuid:{0}", cuuid);
		return;
	}

	clients[cuuid]->forward_hub_call_client(cuuid, _data_bin);

}

void gatemanager::call_group_client(const std::vector<std::string>& cuuids, const std::string& func, const msgpack11::MsgPack::array& argvs) {
	msgpack11::MsgPack::array event_;
	event_.push_back(func);
	event_.push_back(argvs);
	msgpack11::MsgPack _pack(event_);
	auto data = _pack.dump();
	std::vector<uint8_t> _data_bin;
	_data_bin.resize(data.size());
	memcpy(_data_bin.data(), data.data(), data.size());

	std::vector<std::shared_ptr<directproxy> > directs;
	std::unordered_map<std::shared_ptr<gateproxy>, std::vector<std::string> > gate_clients;
	for (auto cuuid : cuuids) {
		auto it_direct_client = direct_clients.find(cuuid);
		if (it_direct_client != direct_clients.end()) {
			directs.push_back(it_direct_client->second);
			continue;
		}

		auto it_gate = clients.find(cuuid);
		if (it_gate != clients.end()) {
			auto it = gate_clients.find(it_gate->second);
			if (it == gate_clients.end()) {
				gate_clients.insert(std::make_pair(it_gate->second, std::vector<std::string>()));
				it = gate_clients.find(it_gate->second);
			}
			it->second.push_back(cuuid);
		}
	}

	{
		if (!directs.empty()) {
			msgpack11::MsgPack::array _argv_array;
			_argv_array.push_back(_data_bin);

			msgpack11::MsgPack::array event_;
			event_.push_back("hub_call_client_call_client");
			event_.push_back(_argv_array);
			msgpack11::MsgPack _pack(event_);
			auto data = _pack.dump();

			size_t len = data.size();
			size_t datasize = len + 4;
			auto _data = (unsigned char*)malloc(datasize);
			auto _crypt_data = (unsigned char*)malloc(datasize);
			_data[0] = len & 0xff;
			_data[1] = len >> 8 & 0xff;
			_data[2] = len >> 16 & 0xff;
			_data[3] = len >> 24 & 0xff;
			memcpy(&_data[4], data.c_str(), data.size());

			memcpy(_crypt_data, _data, datasize);
			service::channel_encrypt_decrypt_ondata::xor_key_encrypt_decrypt((char*)(&(_crypt_data[4])), len);

			std::vector<std::shared_ptr<abelkhan::Ichannel> > crypt_chs;
			std::vector<std::shared_ptr<abelkhan::Ichannel> > chs;
			for (auto direct_proxy : directs) {
				if (direct_proxy->_direct_ch->is_xor_key_crypt()) {
					crypt_chs.push_back(direct_proxy->_direct_ch);
				}
				else {
					chs.push_back(direct_proxy->_direct_ch);
				}
			}

#pragma omp parallel for
			for (auto ch : crypt_chs) {
				ch->send((char*)_crypt_data, datasize);
			}

#pragma omp parallel for
			for (auto ch : chs) {
				ch->send((char*)_data, datasize);
			}

			free(_data);
			free(_crypt_data);
		}
	}

#pragma omp parallel for
	for (auto it : gate_clients) {
		it.first->forward_hub_call_group_client(it.second, _data_bin);
	}
}

void gatemanager::call_global_client(const std::string& func, const msgpack11::MsgPack::array& argvs) {
	msgpack11::MsgPack::array event_;
	event_.push_back(func);
	event_.push_back(argvs);
	msgpack11::MsgPack _pack(event_);
	auto data = _pack.dump();
	std::vector<uint8_t> _data_bin;
	_data_bin.resize(data.size());
	memcpy(_data_bin.data(), data.data(), data.size());

	for (auto gate : gates) {
		gate.second->forward_hub_call_global_client(_data_bin);
	}
}

}

File: csharp_component/component/hub/gatemanager.cs
using Abelkhan;
using MsgPack.Serialization;
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Threading;

namespace Hub
{
    public class DirectProxy
    {
        private readonly Abelkhan.hub_call_client_caller _hub_call_client_caller;

        public readonly string _cuuid;
        public readonly Abelkhan.Ichannel _direct_ch;

        public long _timetmp = 0;
        public long _theory_timetmp = 0;

        public DirectProxy(string cuuid_, Abelkhan.Ichannel direct_ch)
        {
            _cuuid = cuuid_;
            _direct_ch = direct_ch;

            _hub_call_client_caller = new Abelkhan.hub_call_client_caller(_direct_ch, Abelkhan.ModuleMgrHandle._modulemng);
        }

        public void call_client(byte[] rpc_argv)
        {
            _hub_call_client_caller.call_client(rpc_argv);
        }
    }

    public class GateManager
	{
        public string current_client_uuid;

        private readonly Dictionary<string, GateProxy> clients;

        private readonly Dictionary<string, GateProxy> _wait_destory_gateproxys;
        private readonly Dictionary<Abelkhan.Ichannel, GateProxy> ch_gateproxys;
        private readonly Dictionary<string, GateProxy> gates;

        private readonly Dictionary<string, DirectProxy> direct_clients;
        private readonly Dictionary<Abelkhan.Ichannel, DirectProxy> ch_direct_clients;

        private readonly MessagePackSerializer<ArrayList> _serializer = MessagePackSerializer.Get<ArrayList>();

        private readonly Abelkhan.RedisMQ _gate_redismq_conn;

        public GateManager(Abelkhan.RedisMQ _conn)
        {
            _gate_redismq_conn = _conn;

            clients = new Dictionary<string, GateProxy>();

            _wait_destory_gateproxys = new Dictionary<string, GateProxy>();
            ch_gateproxys = new Dictionary<Abelkhan.Ichannel, GateProxy>();
            gates = new Dictionary<string, GateProxy>();

            direct_clients = new Dictionary<string, DirectProxy>();
            ch_direct_clients = new Dictionary<Abelkhan.Ichannel, DirectProxy>();

            Hub._timer.addticktime(10000, heartbeat_tick_hub_health);
            Hub._timer.addticktime(10000, heartbeat_client);
        }

        public void connect_gate(String name)
        {
            var ch = _gate_redismq_conn.connect(name);
            if (ch != null)
            {
                var _proxy = new GateProxy(ch, name);

                if (gates.TryGetValue(name, out GateProxy _old_proxy))
                {
                    _wait_destory_gateproxys.Add(name, _old_proxy);
                    gates[name] = _proxy;
                }
                else
                {
                    gates.Add(name, _proxy);
                }

                ch_gateproxys.Add(ch, _proxy);
                Log.Log.info("connect gate:{0}", name);

                lock (Hub.add_chs)
                {
                    Hub.add_chs.Add(ch);
                }
                _proxy.reg_hub();
            }
        }

        public bool get_gateproxy(Abelkhan.Ichannel gate_ch, out GateProxy proxy)
		{
			return ch_gateproxys.TryGetValue(gate_ch, out proxy);
		}

        public bool get_gateproxy(string session_uuid, out GateProxy proxy)
        {
            return clients.TryGetValue(session_uuid, out proxy);
        }

        public string get_client_gate_name(string session_uuid)
        {
            clients.TryGetValue(session_uuid, out GateProxy _client_gate_proxy);
            if (_client_gate_proxy != null)
            {
                return _client_gate_proxy._name;
            }
            return null;
        }

        public event Action<string> on_gate_closed;
        public void gate_be_closed(string svr_name)
        {
            Log.Log.info("gate_be_closed name:{0}", svr_name);
            if (_wait_destory_gateproxys.Remove(svr_name, out GateProxy _old_proxy))
            {
                var remove = new List<string>();
                foreach (var it in clients)
                {
                    if (it.Value == _old_proxy)
                    {
                        remove.Add(it.Key);
                    }
                }
                foreach (var it in remove)
                {
                    clients.Remove(it);
                }

                ch_gateproxys.Remove(_old_proxy._ch);

                lock (Hub.remove_chs)
                {
                    Hub.remove_chs.Add(_old_proxy._ch);
                }
            }
            else
            {
                if (gates.Remove(svr_name, out GateProxy _proxy))
                {
                    var remove = new List<string>();
                    foreach (var it in clients)
                    {
                        if (it.Value == _proxy)
                        {
                            remove.Add(it.Key);
                        }
                    }
                    foreach (var it in remove)
                    {
                        clients.Remove(it);
                    }

                    ch_gateproxys.Remove(_proxy._ch);

                    lock (Hub.remove_chs)
                    {
                        Hub.remove_chs.Add(_proxy._ch);
                    }

                    on_gate_closed?.Invoke(svr_name);
                }
            }
        }

        public void client_connect(string client_uuid, Abelkhan.Ichannel gate_ch)
        {
            if (ch_gateproxys.TryGetValue(gate_ch, out GateProxy _proxy))
            {
                if (!clients.ContainsKey(client_uuid))
                {
                    clients.Add(client_uuid, _proxy);
                }
            }
            else
            {
                Log.Log.err("invaild gate:{0}, ch_gateproxys.count:{1}", gate_ch, ch_gateproxys.Count);
            }

            Log.Log.trace("client {0} connected", client_uuid);
        }

        public GateProxy client_seep(string client_uuid, string gate_name)
        {
            if (!gates.TryGetValue(gate_name, out GateProxy _proxy))
            {
                Log.Log.err("invaild gate name:{0}", gate_name);
                return null;
            }
            if (!clients.ContainsKey(client_uuid))
            {
                clients.Add(client_uuid, _proxy);
            }

            Log.Log.trace("client {0} seep!", client_uuid);

            return _proxy;
        }

        public event Action<string> clientDisconnect;
        public void client_disconnect(String client_uuid)
        {
            if (clients.Remove(client_uuid))
            {
                clientDisconnect?.Invoke(client_uuid);
            }
        }

        public event Action<string> clientException;
        public void client_exception(String client_uuid)
        {
            clientException?.Invoke(client_uuid);
        }

        public void direct_client_connect(String client_uuid, Abelkhan.Ichannel direct_ch)
        {
            if (direct_clients.Remove(client_uuid, out DirectProxy _directproxy))
            {
                ch_direct_clients.Remove(_directproxy._direct_ch);
            }

            Log.Log.trace("reg direct client:{0}", client_uuid);

            var _directproxy_new = new DirectProxy(client_uuid, direct_ch);
            direct_clients.Add(client_uuid, _directproxy_new);

            ch_direct_clients.Add(direct_ch, _directproxy_new);
        }

        public event Action<string> directClientDisconnect;
        public void direct_client_disconnect(Abelkhan.Ichannel direct_ch)
        {
            if (ch_direct_clients.Remove(direct_ch, out DirectProxy _proxy))
            {
                direct_clients.Remove(_proxy._cuuid);
                directClientDisconnect?.Invoke(_proxy._cuuid);
            }
        }

        public void direct_client_exception(Abelkhan.Ichannel direct_ch)
        {
            if (ch_direct_clients.TryGetValue(direct_ch, out DirectProxy _proxy))
            {
                clientException?.Invoke(_proxy._cuuid);
            }
        }

        void heartbeat_tick_hub_health(long ticktime)
        {
            foreach(var (_, proxy) in ch_gateproxys)
            {
                proxy.tick_hub_health();
            }

            Hub._timer.addticktime(10000, heartbeat_tick_hub_health);
        }

        void heartbeat_client(long ticktime)
        {
            List<DirectProxy> remove_client = new List<DirectProxy>();
            List<DirectProxy> exception_client = new List<DirectProxy>();
            foreach (var item in direct_clients)
            {
                var proxy = item.Value;
                if (proxy._timetmp > 0 && (proxy._timetmp + 10000) < ticktime)
                {
                    remove_client.Add(proxy);
                }
                if (proxy._timetmp > 0 && proxy._theory_timetmp > 0 && (proxy._theory_timetmp - proxy._timetmp) > 10000)
                {
                    exception_client.Add(proxy);
                }
            }

            foreach (var _client in remove_client)
            {
                lock (Hub.remove_chs)
                {
                    Hub.remove_chs.Add(_client._direct_ch);
                }
                direct_client_disconnect(_client._direct_ch);
            }

            foreach (var _client in exception_client)
            {
                direct_client_exception(_client._direct_ch);
            }

            Hub._timer.addticktime(10000, heartbeat_client);
        }

        public bool get_directproxy(Abelkhan.Ichannel direct_ch, out DirectProxy _proxy)
        {
            return ch_direct_clients.TryGetValue(direct_ch, out _proxy);
        }

        public void disconnect_client(String uuid)
        {
            if (clients.Remove(uuid, out GateProxy _proxy))
            {
                _proxy.disconnect_client(uuid);
            }

            if (direct_clients.Remove(uuid, out DirectProxy _directproxy)) {
                ch_direct_clients.Remove(_directproxy._direct_ch);
            }
        }

        public void call_client(String uuid, String func, ArrayList _argvs_list)
		{
            using var st = MemoryStreamPool.mstMgr.GetStream();
            ArrayList _event = new ArrayList
            {
                func,
                _argvs_list
            };
            _serializer.Pack(st, _event);
            st.Position = 0;
            var _rpc_bin = st.ToArray();

            if (direct_clients.TryGetValue(uuid, out DirectProxy _client))
            {
                _client.call_client(_rpc_bin);
            }
            else
            {
                if (clients.ContainsKey(uuid))
                {
                    clients[uuid].forward_hub_call_client(uuid, _rpc_bin);
                }
                else
                {
                    Log.Log.trace("no-exist client:", uuid);
                }
            }
        }

        public void call_group_client(List<string> uuids, String func, ArrayList _argvs_list)
        {
            var _direct_clients = new List<Abelkhan.Ichannel>();
            var _direct_clients_crypt = new List<Abelkhan.Ichannel>();
            var tmp_gates = new Dictionary<GateProxy, List<string> >();
            foreach (var _uuid in uuids)
            {
                if (direct_clients.TryGetValue(_uuid, out DirectProxy _client))
                {
                    if (_client._direct_ch.is_xor_key_crypt())
                    {
                        _direct_clients_crypt.Add(_client._direct_ch);
                    }
                    else
                    {
                        _direct_clients.Add(_client._direct_ch);
                    }
                    continue;
                }

                if (clients.TryGetValue(_uuid, out GateProxy _proxy))
                {
                    if (!tmp_gates.ContainsKey(_proxy))
                    {
                        tmp_gates.Add(_proxy, new List<string>());
                    }
                    tmp_gates[_proxy].Add(_uuid);
                }
            }

            using var st = MemoryStreamPool.mstMgr.GetStream();
            ArrayList _rpc_argv = new ArrayList
            {
                func,
                _argvs_list
            };
            _serializer.Pack(st, _rpc_argv);
            st.Position = 0;
            var _rpc_bin = st.ToArray();

            foreach (var _proxy in tmp_gates)
            {
                _proxy.Key.forward_hub_call_group_client(_proxy.Value, _rpc_bin);
            }

            if (_direct_clients.Count > 0)
            {
                using var st_event = MemoryStreamPool.mstMgr.GetStream();
                var _direct_rpc_argv = new ArrayList
                {
                    _rpc_bin
                };
                ArrayList _event = new ArrayList
                {
                    "hub_call_client_call_client",
                    _direct_rpc_argv
                };
                _serializer.Pack(st_event, _event);
                st_event.Position = 0;
                var data = st_event.ToArray();

                using var st_send = MemoryStreamPool.mstMgr.GetStream();
                var _tmplenght = data.Length;
                st_send.WriteByte((byte)(_tmplenght & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 8) & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 16) & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 24) & 0xff));
                st_send.Write(data, 0, _tmplenght);
                st_send.Position = 0;
                var buf = st_send.ToArray();

                foreach (var _client in _direct_clients)
                {
                    _client.send(buf);
                }
            }

            if (_direct_clients_crypt.Count > 0)
            {
                using var st_event = MemoryStreamPool.mstMgr.GetStream();
                var _direct_rpc_argv = new ArrayList
                {
                    _rpc_bin
                };
                ArrayList _event = new ArrayList
                {
                    "hub_call_client_call_client",
                    _direct_rpc_argv
                };
                _serializer.Pack(st_event, _event);
                st_event.Position = 0;
                var data = st_event.ToArray();

                using var st_send = MemoryStreamPool.mstMgr.GetStream();
                var _tmplenght = data.Length;
                st_send.WriteByte((byte)(_tmplenght & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 8) & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 16) & 0xff));
                st_send.WriteByte((byte)((_tmplenght >> 24) & 0xff));
                st_send.Write(data, 0, _tmplenght);
                st_send.Position = 0;
                var buf = st_send.ToArray();

                using var st_send_crypt = MemoryStreamPool.mstMgr.GetStream();
                st_send_crypt.Write(st_send.GetBuffer());
                st_send_crypt.Position = 0;
                var crypt_buf = st_send_crypt.ToArray();
                Abelkhan.Crypt.crypt_func_send(crypt_buf);

                foreach (var _client in _direct_clients_crypt)
                {
                    _client.send(crypt_buf);
                }
            }
        }

		public void call_global_client(String func, ArrayList _argvs_list)
		{
            var st = MemoryStreamPool.mstMgr.GetStream();
            ArrayList _event = new ArrayList
            {
                func,
                _argvs_list
            };
            _serializer.Pack(st, _event);
            st.Position = 0;
            var _rpc_bin = st.ToArray();
            foreach (var _proxy in ch_gateproxys)
			{
				_proxy.Value.forward_hub_call_global_client(_rpc_bin);
			}
		}
	}
}

